index.js

```diff
@@ -14,255 +12,246 @@ var timedOut = require('timed-out');
+				url: url,
+				method: opts.method,
+				code: err.code
+			}));
+		});
 
-	if (body) {
-		if (typeof body !== 'string' && !Buffer.isBuffer(body) && !isStream.readable(body)) {
-			throw new GotError('options.body must be a ReadableStream, string or Buffer');
+		if (opts.timeout) {
+			timedOut(req, opts.timeout);
 		}
 
-		opts.method = opts.method || 'POST';
-
-		if (!opts.headers['content-length'] && !opts.headers['transfer-encoding'] && !isStream.readable(body)) {
-			var length = typeof body === 'string' ? Buffer.byteLength(body) : body.length;
-			opts.headers['content-length'] = length;
-		}
+		setImmediate(ee.emit.bind(ee), 'request', req);
 	}
 
-	opts.method = opts.method || 'GET';
+	get(opts);
+	return ee;
+}
 
-	// returns a proxy stream to the response
-	// if no callback has been provided
-	if (!cb) {
-		proxy = duplexify();
+function asCallback(opts, cb) {
+	var ee = requestAsEventEmitter(opts);
+	var url = urlLib.format(opts);
 
-		// forward errors on the stream
-		cb = function (err, data, response) {
-			proxy.emit('error', err, data, response);
-		};
-	}
-
-	if (proxy && json) {
-		throw new GotError('got can not be used as stream when options.json is used');
-	}
+	ee.on('request', function (req) {
+		if (isStream.readable(opts.body)) {
+			opts.body.pipe(req);
+			opts.body = undefined;
+			return;
+		}
 
-	function get(opts, cb) {
-		var fn = opts.protocol === 'https:' ? https : http;
-		var url = urlLib.format(opts);
+		req.end(opts.body);
+	});
 
-		if (opts.agent === undefined) {
-			opts.agent = infinityAgent[fn === https ? 'https' : 'http'].globalAgent;
-
-			if (process.version.indexOf('v0.10') === 0 && fn === https && (
-				typeof opts.ca !== 'undefined' ||
-				typeof opts.cert !== 'undefined' ||
-				typeof opts.ciphers !== 'undefined' ||
-				typeof opts.key !== 'undefined' ||
-				typeof opts.passphrase !== 'undefined' ||
-				typeof opts.pfx !== 'undefined' ||
-				typeof opts.rejectUnauthorized !== 'undefined')) {
-				opts.agent = new infinityAgent.https.Agent({
-					ca: opts.ca,
-					cert: opts.cert,
-					ciphers: opts.ciphers,
-					key: opts.key,
-					passphrase: opts.passphrase,
-					pfx: opts.pfx,
-					rejectUnauthorized: opts.rejectUnauthorized
-				});
+	ee.on('response', function (res) {
+		readAllStream(res, opts.encoding, function (err, data) {
+			if (err) {
+				cb(new got.ReadError(err.message, {url: url}), null, res);
+				return;
 			}
-		}
 
-		var req = fn.request(opts, function (response) {
-			var statusCode = response.statusCode;
-			var res = response;
+			var statusCode = res.statusCode;
 
-			// auto-redirect only for GET and HEAD methods
-			if (isRedirect(statusCode) && 'location' in res.headers && (opts.method === 'GET' || opts.method === 'HEAD')) {
-				// discard response
-				res.resume();
+			if (statusCode < 200 || statusCode > 299) {
+				err = new got.HTTPError({statusCode: statusCode, url: url, method: opts.method});
```

Comment:
You pass in `url` and `method` here, but I don't see it used in the error construction.